include/fiddle/interaction/interaction_utilities.h

```diff
@@ -260,6 +263,155 @@ namespace fdl
                        const Vector<double>         &position,
                        const Vector<double>         &spread_values);
 
+  /**
+   * Compute intersection the point of a line with an edge
+   *
+   * WARNING: This code is specialized to the case in which q is a unit vector
+   * aligned with the coordinate axes.
+   *
+   * Linear interpolation:
+   *
+   * 0.5*(1-u)*p0 + 0.5*(1+u)*p1 = r + t*q
+   *
+   * Factor the interpolation formula:
+   *
+   * 0.5*(p1-p0)*u+0.5*(p1+p0) = r + t*q
+   *
+   * Solve for u:
+   *
+   *    a*u + b = 0
+   *
+   * with:
+   *
+   *   a = 0.5*(-p0+p1)
+   *   b = 0.5*(p0+p1) - r
+   * @param[out] t_vals The vector containing the intersection point in Eulerian
+   * coordinate and element reference coordinate.
+   *
+   * @param[in] element_vertices The array containing Eulerian location of
+   * vertices of the given element.
+   *
+   * @param[in] r The center location of the FD stencil`.
+   *
+   * @param[in] q a unit vector aligned with the coordinate axes.
+   *
+   * @param[in] tol a given tolerance for deciding if the intersection point is
+   * within the edge.
+   *
+   * @note This function is only work for P1 element.
+   */
+  void
+  intersect_line_with_edge(std::vector<std::pair<double, Point<1>>> &t_vals,
+                           std::array<Point<2>, 2> element_vertices,
+                           dealii::Point<2>        r,
+                           dealii::Tensor<1, 2>    q,
+                           const double            tol);
```

Comment:
Lets improve the API vs. what is in IBAMR:
```cpp
template <int spacedim>
std_cxx17::optional<double>
intersect_stencil_with_simplex(const std::array<Point<spacedim>, spacedim + 1> &simplex,
                               const Point<spacedim>                           &stencil_start,
                               const double                                    &stencil_width,
                               const unsigned int                               stencil_axis);
```
If it intersects we populate the [optional](https://en.cppreference.com/w/cpp/utility/optional) object: if not then it is empty. We need to avoid dynamic memory allocation for performance reasons and this better documents the "zero or one" nature.

We are going to handle checking for duplicates in a different way (that's what IntersectionPatchMap is for: we find all of them on a patch and uniquify) so we don't need `tol` any more.

Reply:
Here the return value should be the convex combination coefficient between the start of the FD stencil and its end - see also https://github.com/drwells/fiddle/blob/master/include/fiddle/grid/patch_intersection_map.h#L62 which defines intersections in that way.